### PR TITLE
#65 Displays Loading overlay message when pss service is running

### PR DESCRIPTION
**Pull request by mohsinshabir1234**

### CODE REVIEW

File: gui/control.js [hunk 1]
```diff
@@ -17,13 +17,45 @@ const addAccountButton = document.getElementById('add-account-btn');
...
+    // Wait for the asynchronous task to complete
+    await toRun();
+  } finally {
+    loadingOverlay.style.display = "none";
+  }
+}
+
+
+async function switchAcc(nickname) {
+  await runWithLoading(async () => {
```
Comment (adsau59) [inline gui/control.js:41]: dont see the need to have switchAcc an async function

File: gui/control.js [hunk 2]
```diff
@@ -17,13 +17,45 @@ const addAccountButton = document.getElementById('add-account-btn');
...
+  logger.debug(`switching ${nickname}`)
+  const error = await PSS.switchAccount(nickname);
+  if (!error) return;
+  showToast("Error!", errorToMessage[error])
+});
```
Comment (adsau59) [inline gui/control.js:46]: correct indentation

File: gui/control.js [hunk 3]
```diff
@@ -17,13 +17,45 @@ const addAccountButton = document.getElementById('add-account-btn');
...
 async function addButtonPressed() {
+ 
+    await runWithLoading(async () => {
     const nickname = await showTextInputPopup("Enter Nickname", "Add Account");
     if (!nickname) return;
     const error = await PSS.addAccount(nickname);
     if (!error) return;
-    showToast("Error!", errorToMessage[error])
+    showToast("Error!", errorToMessage[error]);
+  });
```
Comment (adsau59) [inline gui/control.js:57]: correct indentation

File: gui/control.js [hunk 4]
```diff
@@ -17,13 +17,45 @@ const addAccountButton = document.getElementById('add-account-btn');
...
+  logger.debug(`switching ${nickname}`)
+  const error = await PSS.switchAccount(nickname);
+  if (!error) return;
+  showToast("Error!", errorToMessage[error])
+});
+}
 
 async function addButtonPressed() {
+ 
+    await runWithLoading(async () => {
```
Comment (adsau59) [inline gui/control.js:51]: again, I dont see why addButtonPressed is async

File: gui/control.js [hunk 5]
```diff
@@ -45,12 +77,9 @@ function render() {
...
         // Deleting User Profile
-        document.getElementById(`switch-btn-${nickname}`).addEventListener('click', async function () {
-            logger.debug(`switching ${nickname}`)
-            const error = await PSS.switchAccount(nickname);
-            if (!error) return;
-            showToast("Error!", errorToMessage[error])
-        });
+        document.getElementById(`switch-btn-${nickname}`).addEventListener('click',()=>{
+            switchAcc(nickname);
+        });
```
Comment (adsau59) [inline gui/control.js:82]: no need for curly braces if arrow function is of 1 line

```js
document.getElementById(`switch-btn-${nickname}`).addEventListener('click',()=switchAcc(nickname)); 
```

File: gui/control.js [hunk 6]
```diff
@@ -64,7 +93,8 @@ function render() {
 }
 
 async function main() {
-
+ 
+   runWithLoading(async () => {
```
Comment (adsau59) [inline gui/control.js:97]: first render doesn't take much time, we dont need to show loading for that

**Review by adsau59 — CHANGES_REQUESTED** [review]
requested changes, apart from that looks good

**Comment by mohsinshabir1234** [comment]
#65

File: gui/control.js [hunk 7]
```diff
@@ -17,13 +17,44 @@ const addAccountButton = document.getElementById('add-account-btn');
...
 window.addEventListener('DOMContentLoaded', main)
 
+const loadingMessage = document.getElementById('loading-message');
+const loadingOverlay = document.getElementById('loading-overlay');
+const showOverlay =false;
 
-async function addButtonPressed() {
+function setLoadingOverlayVisibility(showOverlay) {
+  loadingOverlay.style.display = showOverlay ? 'flex' : 'none';
+}
```
Comment (adsau59) [inline gui/control.js:26]: if this is no longer needed, pls remove it

File: gui/control.js [hunk 8]
```diff
@@ -17,13 +17,44 @@ const addAccountButton = document.getElementById('add-account-btn');
...
+async function runWithLoading(toRun) {
+  try {
+    loadingOverlay.style.display = "flex";
+
+    // Wait for the asynchronous task to complete
+    await toRun();
+  } finally {
+    loadingOverlay.style.display = "none";
+  }
+}
```
Comment (adsau59) [inline gui/control.js:37]: move this to elements.js

File: gui/control.js [hunk 9]
```diff
@@ -17,13 +17,44 @@ const addAccountButton = document.getElementById('add-account-btn');
 window.addEventListener('unload', beforeQuit)
 window.addEventListener('DOMContentLoaded', main)
 
+const loadingMessage = document.getElementById('loading-message');
+const loadingOverlay = document.getElementById('loading-overlay');
```
Comment (adsau59) [inline gui/control.js:21]: move this inside runWithLoading function

File: gui/control.js [hunk 10]
```diff
@@ -17,13 +17,44 @@ const addAccountButton = document.getElementById('add-account-btn');
...
+
+    // Wait for the asynchronous task to complete
+    await toRun();
+  } finally {
+    loadingOverlay.style.display = "none";
+  }
+}
+
+function switchAcc(nickname) {
+  await runWithLoading(async () => {
```
Comment (adsau59) [inline gui/control.js:40]: will need ot remove await from this line

File: gui/control.js [hunk 11]
```diff
@@ -17,13 +17,44 @@ const addAccountButton = document.getElementById('add-account-btn');
...
+    logger.debug(`switching ${nickname}`)
+    const error = await PSS.switchAccount(nickname);
+    if (!error) return;
+    showToast("Error!", errorToMessage[error])
+  });
+}
+
+function addButtonPressed() {
+
+  await runWithLoading(async () => {
```
Comment (adsau59) [inline gui/control.js:50]: will need ot remove await from this line

File: gui/style.css [hunk 12]
```diff
@@ -199,3 +199,30 @@ input{
     bottom: 10px;
     right: 10px;
 }
+#loading-overlay {
```
Comment (adsau59) [inline gui/style.css:202]: add a new line above this for clean code

**Review by adsau59 — CHANGES_REQUESTED** [review]
changes requested

File: gui/control.js (hunk 7 above)
Comment (mohsinshabir1234) [reply gui/control.js:26]: Removed

File: gui/control.js (hunk 8 above)
Comment (mohsinshabir1234) [reply gui/control.js:37]: The loading function doesnt work after moving to element.js

File: gui/control.js (hunk 9 above)
Comment (mohsinshabir1234) [reply gui/control.js:21]: Have removed loading message as there was no need for it. Moved loading overlay message inside loadingOverlay.

File: gui/control.js (hunk 10 above)
Comment (mohsinshabir1234) [reply gui/control.js:40]: Have already done this in previous commit.

File: gui/control.js (hunk 11 above)
Comment (mohsinshabir1234) [reply gui/control.js:50]: Have done it in previous commit.

File: gui/style.css (hunk 12 above)
Comment (mohsinshabir1234) [reply gui/style.css:202]: Done

File: gui/control.js [hunk 13]
```diff
@@ -65,16 +86,18 @@ function render() {
...
+  logger.debug("initializing")
+  if (errorCode) showToast("Error!", errorToMessage[errorCode]);
+  if (!global_state.flags.parsecDataLocationFound) showToast("Error!", errorToMessage[error.PARSEC_NOT_INSTALLED])
+  if (!global_state.flags.parsecdFound) showToast("Error!", error.PARSECD_NOT_IN_DEFAULT);
+
+  global_state.onConfigChanged.push(render);
+  addAccountButton.addEventListener('click', addButtonPressed);
+  render();
+
+}
```
Comment (adsau59) [inline gui/control.js:100]: pls make sure not to commit changes u didn't change

File: gui/control.js (hunk 8 above)
Comment (adsau59) [reply gui/control.js]: why not?

**Review by adsau59 — CHANGES_REQUESTED** [review]
changes requested

File: gui/control.js (hunk 8 above)
Comment (mohsinshabir1234) [reply gui/control.js]: Done

**Review by mohsinshabir1234 — COMMENTED** [review]
done with the required changes

File: gui/control.js (hunk 13 above)
Comment (mohsinshabir1234) [reply gui/control.js:87]: done

**Comment by adsau59** [comment]
merged, thanks for contributing to the project

**Pull request merged**